pkg/frontend/querymiddleware/split_and_cache.go

```diff
@@ -380,7 +383,15 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
 }
 
 // storeCacheExtents stores the extents for given key in the cache.
-func (s *splitAndCacheMiddleware) storeCacheExtents(ctx context.Context, key string, extents []Extent) {
+func (s *splitAndCacheMiddleware) storeCacheExtents(ctx context.Context, key string, tenantIDs []string, extents []Extent) {
+	lowerTTL := false
+	lowerTTLWithinTimePeriod := validation.MaxDurationPerTenant(tenantIDs, func(tenantID string) time.Duration {
+		return time.Duration(s.limits.OutOfOrderTimeWindow(tenantID))
+	})
+	if lowerTTLWithinTimePeriod > 0 && len(extents) > 0 {
+		lowerTTL = extents[len(extents)-1].End >= time.Now().Add(-lowerTTLWithinTimePeriod).UnixMilli()
```

Comment:
From a very quick look at the code, I can't get if it's always guaranteed extends are sorted by end time. Have you found a proof of it? Otherwise it's probably safer to iterate through all extents.